### PR TITLE
feat(typescript-estree): add maximumDefaultProjectFileMatchCount and wide allowDefaultProjectForFiles glob restrictions

### LINKED ISSUE #8923: Enhancement: ban wide globs in `parserOptions.useProjectService.allowDefaultProjectForFiles`
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

We should lock down this API to start with and widen it as needed.
In particular we know that the default project can and will cause perf issues if used incorrectly.
EG if it matches too many files it'll cause dramatic slowdowns.

Based on what @JoshuaKGoldberg has found - it'll create one new `ts.Program` per file. This process is slow - but given the number of files is small then it shouldn't impact things too badly. If someone does something dumb like pass a glob like `**/*.ts` to match their entire project then their run will be stupidly slow for no good reason.

In a nutshell - the motivation against this was the entire reason we moved away from "default programs" back in [v2.0.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0). Because people accidentally fell into the default program and their lint runs slowed to a crawl because of them.

We should do the following:

(1) ban obviously dumb globs like `*`, `**/*`, `**/*.*`, etc. Essentially any glob that doesn't match the regex `\w` should be banned with an explicit failure like

> Unnecessarily wide globs are disallowed. If you have a usecase for them please file an issue in the typescript-eslint project.

(2) ban too many files matching the glob. We should pick a sensible default that we think won't noticably impact the run performance (say, 5 files) and if the lint run matches against more than that many files - we should explicitly fail like

> `{count}` files have matched the default project. Having many files run with the default project is known to cause performance issues and slow down your lint run.
> Consider creating a separate `tsconfig.json` to match these files `https://typescript-eslint.io/docs/link/to/guide/about/this/optimisation`.
> If you're ok with slower lint runs then set `parserOptions.iLikeSlowLintRunsAndICannotLie_DO_NOT_USE_THIS_OR_YOU_WILL_BE_FIRED` to an integer greater than `{count}`

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8923
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds the two safety checks for `allowDefaultProjectForFiles` as suggested in the issue:
* A new `maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING` option with a default value of `8`
* A check that `allowDefaultProjectForFiles` doesn't include `**`
   * Intentionally allows slightly more permissive matches like `./*.*s`, as that's what I'd expect to be convenient & reasonable alongside the count maximum

Adds an FAQs blog post to explain them.

💖

### CODE REVIEW

File: docs/troubleshooting/FAQ.mdx [hunk 1]
```diff
@@ -499,6 +533,6 @@ If you think you're having issues with performance, see our [Performance Trouble
 
 ## Are TypeScript project references supported?
 
-No, TypeScript project references are not yet supported.
+Yes, but only with [`EXPERIMENTAL_useProjectService`](../packages/Parser.mdx#experimental_useprojectservice).
```
Comment (JoshuaKGoldberg) [inline docs/troubleshooting/FAQ.mdx:536]: Aside: we'll definitely want more explanation here 😄 but quickly noted this in the meantime. Once this & a few more v8 milestone PRs are in, #7350 tracks a full blog post / explainer / revamp.

**Review by bradzacher — APPROVED** [review]

**Comment by me4502** [comment]
Is this meant to impact situations where `EXPERIMENTAL_useProjectService` is set to `true`? Not sure if it's a bug (will make an issue if it is), but I'm now seeing `Parsing error: Too many files (>8) have matched the default project.` for every single file in each workspaces' lint run as of 7.8.0.

From my understanding the project service should be loading the closest `tsconfig.json` in the file tree, and using the default if it doesn't match an included file, correct? It seems to be either ignoring the tsconfig.json file per workspace or ignoring the `includes` values in each one.

**Comment by JoshuaKGoldberg** [comment]
Hmm, every single file hitting this does sound like a bug. I'd appreciate an issue - yes please!

**Comment by me4502** [comment]
Will do, thanks. I haven't managed to reproduce this outside of one specific large & complex monorepo, so I'll see if I can narrow it down to reproduce.

**Comment by ehoogeveen-medweb** [comment]
I'm also seeing this match every file when linting a directory containing a project config, even after setting `allowDefaultProjectForFiles: []`. I'll try to get a minimal reproduction.

**Comment by stepankuzmin** [comment]
Having the same issue with `Too many files (>8) have matched the default project` after upgrading to `7.8.0`.

**Comment by JoshuaKGoldberg** [comment]
Heyall, per our [PR contributing guide](https://typescript-eslint.io/contributing/pull-requests) please don't post on resolved PRs. The comments tend to get lost and this isn't an effective way to discuss issues. Please [file a new issue](https://github.com/typescript-eslint/typescript-eslint/issues/new/choose) with the appropriate template if you've got something to report that hasn't yet been reported.

#9032 tracks the general case of this error. Nobody's investigated to see what the actual issue is (if you have time, please do!). That would be the issue to subscribe to for updates. If you've got any _new information_ to add such as investigation results or a new kind of reproduction, please add it there. Saying _"I also have this issue!"_ doesn't give us any information and just gives an unnecessary notification to everyone in the thread.

`maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING` is appropriately obnoxiously named. It's a "bandaid" solution to files being unmatched by the project service. A better solution is to figure out why they're unmatched and fix that bug - which is either in the ESLint config being used or in typescript-eslint.

**Pull request merged**